### PR TITLE
Fix #1892: Format FractionInputIsEquivalentToRuleClassifier

### LINKED ISSUE #1892: Thoroughly test FractionInputIsEquivalentToRuleClassifier
Since the FractionInputIsEquivalentToRuleClassifier was copied from Oppia web, it didn't have corresponding tests added to verify correctness. These should be ported over from Oppia web as well to ensure that each classifier has working functionality.

Useful reference - https://github.com/oppia/oppia-android/tree/develop/domain/src/test/java/org/oppia/android/domain/classify/rules

### PR DESCRIPTION
**Pull request by peculiaruc**
## Explanation

- Fixes #1892: Format FractionInputIsEquivalentToRuleClassifie.

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Comment by peculiaruc** [comment]
@rt4914 PTAL

**Comment by peculiaruc** [comment]
@rt4914

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+
+  private val WHOLE_NUMBER_VALUE_123 =
+    InteractionObjectTestBuilder.createWholeNumber(isNegative = false, value = 123)
+  private val WHOLE_NUMBER_VALUE_254 =
+    InteractionObjectTestBuilder.createWholeNumber(isNegative = false, value = 254)
+  private val FRACTION_VALUE_2_OVER_8 =
+    InteractionObjectTestBuilder.createFraction(isNegative = false, numerator = 2, denominator = 8)
+  private val FRACTION_VALUE_1_OVER_5 =
+    InteractionObjectTestBuilder.createFraction(isNegative = false, numerator = 1, denominator = 5)
+  private val FRACTION_VALUE_33_OVER_2 = InteractionObjectTestBuilder.createFraction(
```
Comment (rt4914) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: `FRACTION_VALUE_33_OVER_2 ` but the numerator is `133`

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+      Fraction.newBuilder()
+        .setIsNegative(isNegative)
+        .setWholeNumber(wholeNumber)
+        .setNumerator(numerator)
+        .setDenominator(denominator)
+        .build()
+    ).build()
+  }
+
+  private fun createNonNegativeInt(value: Int): InteractionObject {
```
Comment (rt4914) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: This function is already available in `InteractionObjectTestBuilder`

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+      Fraction.newBuilder()
+        .setIsNegative(isNegative)
+        .setWholeNumber(value)
+        .setNumerator(0)
+        .setDenominator(1)
+        .build()
+    ).build()
+  }
+
+  private fun createMixedNumber(
```
Comment (rt4914) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: This function is already available in `InteractionObjectTestBuilder`

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+    return InteractionObject.newBuilder().setFraction(
+      Fraction.newBuilder()
+        .setIsNegative(isNegative)
+        .setNumerator(numerator)
+        .setDenominator(denominator)
+        .build()
+    ).build()
+  }
+
+  private fun createWholeNumber(isNegative: Boolean, value: Int): InteractionObject {
```
Comment (rt4914) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: This function is already available in `InteractionObjectTestBuilder`

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+    }
+
+    assertThat(exception)
+      .hasMessageThat()
+      .contains(
+        "Expected input value to be of type FRACTION not NON_NEGATIVE_INT"
+      )
+  }
+
+  private fun createFraction(
```
Comment (rt4914) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: This function is already available in `InteractionObjectTestBuilder`

**Review by rt4914 — CHANGES_REQUESTED** [review]
I have suggested few changes.

Also, can you post a screenshot showing that tests are passing locally for you.

Thanks @peculiaruc

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt (hunk 1 above)
Comment (peculiaruc) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: Done

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt (hunk 5 above)
Comment (peculiaruc) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: Yes it is in InteractionObjectTestBuilder

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt (hunk 4 above)
Comment (peculiaruc) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: Yes it is in InteractionObjectTestBuilder

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt (hunk 3 above)
Comment (peculiaruc) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: Yes it is in InteractionObjectTestBuilder

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt (hunk 2 above)
Comment (peculiaruc) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: Yes it is in InteractionObjectTestBuilder

**Comment by peculiaruc** [comment]
Screenshot from 2020-10-27 21-34-10

**Comment by peculiaruc** [comment]
@rt4914 PTAL

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt (hunk 2 above)
Comment (anandwana001) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: @peculiaruc As per @rt4914 As the function is already there in `InteractionObjectTestBuilder`, we don't need the same function in this file, so you should remove this.

Comment (peculiaruc) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: @anandwana001  removing this function will make the variable not to have enough information to infer with. I think we should let this one be and remove others. thanks

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt (hunk 3 above)
Comment (peculiaruc) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: Done

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt (hunk 4 above)
Comment (peculiaruc) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: Done

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt (hunk 5 above)
Comment (peculiaruc) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: Done

**Comment by peculiaruc** [comment]
@rt4914 PTAL

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+import kotlin.reflect.KClass
+import kotlin.reflect.full.cast
+import kotlin.test.fail
+
+@RunWith(AndroidJUnit4::class)
+@LooperMode(LooperMode.Mode.PAUSED)
+@Config(manifest = Config.NONE)
+class FractionInputIsEquivalentToRuleClassifierProviderTest {
+
+  private val WHOLE_NUMBER_VALUE_123 =
```
Comment (anandwana001) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: `WHOLE_NUMBER_VALUE_TEST_123 `

Comment (anandwana001) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: applies for others as well.

**Comment by anandwana001** [comment]
Adding @Sarthak2601, as he is the issue author and needs to check if in a bit more detail.

**Review by anandwana001 — CHANGES_REQUESTED** [review]

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt (hunk 6 above)
Comment (peculiaruc) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: Done

**Comment by peculiaruc** [comment]
@rt4914 PTAL

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+        answer = FRACTION_VALUE_TEST_33_OVER_2,
+        inputs = inputs
+      )
+
+    assertThat(matches).isTrue()
+  }
+
+  @Test
+  fun testEquality_fraction2Over133Answer_withFaction1Over242Input_bothValuesNotEquivalent() {
+    val inpus = mapOf("f" to FRACTION_VALUE_TEST_242_OVER_1)
```
Comment (rt4914) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: Typo:
`inpus` -> `inputs`

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+      inputIsEquivalentToRuleClassifier.matches(
+        answer = FRACTION_VALUE_TEST_33_OVER_2,
+        inputs = inpus
+      )
+
+    assertThat(matches).isFalse()
+  }
+
+  @Test
+  fun testEquality_mixedNumbr6And1Over2Answer_withMixedNumber6And1Over2Input_bothValuesMatch() {
```
Comment (rt4914) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: Typo:
`Numbr` -> `Number`

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+      inputIsEquivalentToRuleClassifier.matches(
+        answer = MIXED_NUMBER_VALUE_TEST_6_1_OVER_2,
+        inputs = inputs
+      )
+
+    assertThat(matches).isFalse()
+  }
+
+  @Test
+  fun testEqualuality_missingInputF_throwsException() {
```
Comment (rt4914) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: Typo:
`Equaluality` -> `Equality`

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+      inputIsEquivalentToRuleClassifier.matches(
+        answer = WHOLE_NUMBER_VALUE_TEST_254,
+        inputs = inputs
+      )
+
+    assertThat(matches).isFalse()
+  }
+
+  @Test
+  fun testEquality_negativeMixedNumberAnswer_withPositveMixedNumberInput_bothValuesDoNotMatche() {
```
Comment (rt4914) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: Typo
`Positve` -> `Positive`

**Review by rt4914 — CHANGES_REQUESTED** [review]
@peculiaruc Suggested some typo related changes.

**Comment by peculiaruc** [comment]
@rt4914 PTAL

**Review by rt4914 — APPROVED** [review]
LGTM, thanks.

**Comment by rt4914** [comment]
@peculiaruc Please reply to all the comments that I have made, unless and until we do not resolve all the comments we cannot merge a PR.

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt (hunk 7 above)
Comment (peculiaruc) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: Done

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt (hunk 8 above)
Comment (peculiaruc) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: Done

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt (hunk 9 above)
Comment (peculiaruc) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: Done

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt (hunk 10 above)
Comment (peculiaruc) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: Done

**Comment by peculiaruc** [comment]
> I have suggested few changes.
> 
> Also, can you post a screenshot showing that tests are passing locally for you.
> 
> Thanks @peculiaruc

Noted. Done

**Comment by peculiaruc** [comment]
> @peculiaruc Please reply to all the comments that I have made, unless and until we do not resolve all the comments we cannot merge a PR.

Noted, Done

**Comment by peculiaruc** [comment]
> Adding @Sarthak2601, as he is the issue author and needs to check if in a bit more detail.

Noted

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+
+  private val WHOLE_NUMBER_VALUE_TEST_123 =
+    InteractionObjectTestBuilder.createWholeNumber(isNegative = false, value = 123)
+  private val WHOLE_NUMBER_VALUE_TEST_254 =
+    InteractionObjectTestBuilder.createWholeNumber(isNegative = false, value = 254)
+  private val FRACTION_VALUE_TEST_2_OVER_8 =
+    InteractionObjectTestBuilder.createFraction(isNegative = false, numerator = 2, denominator = 8)
+  private val FRACTION_VALUE_TEST_1_OVER_5 =
+    InteractionObjectTestBuilder.createFraction(isNegative = false, numerator = 1, denominator = 5)
+  private val FRACTION_VALUE_TEST_33_OVER_2 =
```
Comment (anandwana001) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt:48]: keep the consistency in the line break like this one

```kotlin
private val NON_NEGATIVE_VALUE_TEST_679_2_OVER_3 =
    InteractionObjectTestBuilder.createMixedNumber(
      isNegative = true,
      wholeNumber = 679,
      numerator = 2,
      denominator = 3
    )
```

**Review by anandwana001 — CHANGES_REQUESTED** [review]
Remove the TODO from its provider file.

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt (hunk 6 above)
Comment (rt4914) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt]: @anandwana001 Please resolve this comment if possible.

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt (hunk 11 above)
Comment (peculiaruc) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToRuleClassifierProviderTest.kt:48]: Done

**Review by anandwana001 — APPROVED** [review]
LGTM, Thanks @peculiaruc

**Pull request merged**